### DISCUSSION TITLE
Tutorial sqlite connection doesn't seem to work

### DISCUSSION THREAD

**Discussion by IvanDrag0** [lowdefy/lowdefy#1685] [Questions and Help]
I'm trying to go through the [tutorial](https://docs.lowdefy.com/tutorial-start) on the Lowdefy website and I'm stuck at the "[Display ticket data](https://docs.lowdefy.com/tutorial-display-data-page)" step. I did everything per the tutorial (and even copied the files directly from the repo), however, I keep getting an error.

When I'm trying to follow the steps as-is, once I click the submit button, I get the following "String expected" error:

```
{
    "level": 50,
    "time": 1696444639245,
    "name": "lowdefy_server",
    "rid": "889c3662-6541-4277-a063-2a140b56023a",
    "err": {
        "type": "RequestError",
        "message": "String expected",
        "stack": "RequestError: String expected\n    at callRequestResolver (file:///C:/XXXXXXXXXXXXXXX/lowdefy/.lowdefy/dev/node_modules/.pnpm/@lowdefy+api@4.0.0-rc.10/node_modules/@lowdefy/api/dist/routes/request/callRequestResolver.js:28:21)\n    at async callRequest (file:///C:/XXXXXXXXXXXXXXX/lowdefy/.lowdefy/dev/node_modules/.pnpm/@lowdefy+api@4.0.0-rc.10/node_modules/@lowdefy/api/dist/routes/request/callRequest.js:77:22)\n    at async handler (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\.next\\server\\pages\\api\\request\\[pageId]\\[requestId].js:111:22)\n    at async wrappedHandler (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\.next\\server\\chunks\\930.js:245:30)\n    at async Object.apiResolver (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\api-utils\\node.js:366:9)\n    at async NextNodeServer.runApi (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\next-server.js:481:9)\n    at async Object.fn (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\next-server.js:741:37)\n    at async Router.execute (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\router.js:252:36)\n    at async NextNodeServer.run (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\base-server.js:365:29)\n    at async NextNodeServer.handleRequest (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\base-server.js:303:20)",
        "name": "RequestError"
    },
    "user": {},
    "url": "/api/request/new-ticket/save_new_ticket",
    "method": "POST",
    "msg": "String expected"
}

```

I tried adding the database file directly to the lowdefy.yaml (instead of using the .env file) and changed lowdefy.yaml from

```
connections:
  - id: knex
    type: Knex
    properties:
      client: sqlite
      connection:
        filename:
          _secret: SQLITE_FILENAME
```

to

```
connections:
  - id: knex
    type: Knex
    properties:
      client: sqlite
      connection:
        filename: ./tutorial_db.sqlite
```

When I try to submit the form again, I get a new error message:

```
{
    "level": 50,
    "time": 1696444677469,
    "name": "lowdefy_server",
    "rid": "e62c3649-6a02-41cd-a8cb-8466af58aaf4",
    "err": {
        "type": "RequestError",
        "message": "INSERT INTO tickets (\nticket_title,\nticket_type,\nticket_description,\nproduct,\npurchase_in_last_month,\ncreated_date)\nVALUES(\n'dsfsdfs',\n'Suggestion',\n'dfsdfsdf',\n'iPhone 9',\nNULL,\n'2023-10-04 13:37:57.465')\n - SQLITE_ERROR: no such table: tickets",
        "stack": "RequestError: INSERT INTO tickets (\nticket_title,\nticket_type,\nticket_description,\nproduct,\npurchase_in_last_month,\ncreated_date)\nVALUES(\n'dsfsdfs',\n'Suggestion',\n'dfsdfsdf',\n'iPhone 9',\nNULL,\n'2023-10-04 13:37:57.465')\n - SQLITE_ERROR: no such table: tickets\n    at callRequestResolver (file:///C:/XXXXXXXXXXXXXXX/lowdefy/.lowdefy/dev/node_modules/.pnpm/@lowdefy+api@4.0.0-rc.10/node_modules/@lowdefy/api/dist/routes/request/callRequestResolver.js:28:21)\n    at async callRequest (file:///C:/XXXXXXXXXXXXXXX/lowdefy/.lowdefy/dev/node_modules/.pnpm/@lowdefy+api@4.0.0-rc.10/node_modules/@lowdefy/api/dist/routes/request/callRequest.js:77:22)\n    at async handler (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\.next\\server\\pages\\api\\request\\[pageId]\\[requestId].js:111:22)\n    at async wrappedHandler (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\.next\\server\\chunks\\930.js:245:30)\n    at async Object.apiResolver (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\api-utils\\node.js:366:9)\n    at async NextNodeServer.runApi (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\next-server.js:481:9)\n    at async Object.fn (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\next-server.js:741:37)\n    at async Router.execute (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\router.js:252:36)\n    at async NextNodeServer.run (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\base-server.js:365:29)\n    at async NextNodeServer.handleRequest (C:\\XXXXXXXXXXXXXXX\\lowdefy\\.lowdefy\\dev\\node_modules\\.pnpm\\next@12.3.4_react-dom@18.2.0_react@18.2.0\\node_modules\\next\\dist\\server\\base-server.js:303:20)",
        "name": "RequestError"
    },
    "user": {},
    "url": "/api/request/new-ticket/save_new_ticket",
    "method": "POST",
    "msg": "INSERT INTO tickets (\nticket_title,\nticket_type,\nticket_description,\nproduct,\npurchase_in_last_month,\ncreated_date)\nVALUES(\n'dsfsdfs',\n'Suggestion',\n'dfsdfsdf',\n'iPhone 9',\nNULL,\n'2023-10-04 13:37:57.465')\n - SQLITE_ERROR: no such table: tickets"
}
```

However, the "tickets" table is in the database:

![image](https://github.com/lowdefy/lowdefy/assets/24702285/edbc95c1-7ee6-4193-9895-d99a9669062a)

And when I try to run the query manually, it works as expected:

![image](https://github.com/lowdefy/lowdefy/assets/24702285/ffa7605f-e2aa-497a-a3f8-03c1169bc2d5)

![image](https://github.com/lowdefy/lowdefy/assets/24702285/c0b62d45-9062-47e4-b91f-2591d09b7e90)

**Comment by SamTolmay** [comment]
Hi @IvanDrag0 

Thanks for the very detailed issue.

I think the filename property should be the absolute file path, so it would be something like:
```
/Users/user/projects/lowdefy/tutorial/tutorial_db.sqlite
```

If that works then the `String expected` is probably because the secret is not being read correctly- the environment variable should be `LOWDEFY_SECRET_SQLITE_FILENAME`, the `LOWDEFY_SECRET_` gets removed from the secret name.

**Comment by IvanDrag0** [comment]
Hi @SamTolmay,

I just erased the project, copied the one from the example repo, and used the full and relative paths. The full path did work (not sure why it didn't work yesterday), but the relative path still failed. After a bit of testing and digging around, I might have found the reason that the relative paths don't work (at least on Windows).

If I specify a relative path for the sqlite file (the same way I did in my original issue), I get the error that the table doesn't exist. However, that's an incorrect error since if the app can't find the file, it should just say that the file is not found. I believe the reason that it gives an error regarding the table is because it creates a blank (0 Kb) sqlite file in the root of the application itself (not where the yaml configs are).

![image](https://github.com/lowdefy/lowdefy/assets/24702285/4c5d11b5-5160-4a5f-99da-cfee26cc99ee)

![image](https://github.com/lowdefy/lowdefy/assets/24702285/39fc038c-6f50-42e5-a5ac-3b0225ba0055)

Which explains the error since it can actually find the file (since the app seem to create it), but the file is blank. When I stopped the program and replaced the 0 Kb file with the database file I used, everything worked as expected.

![image](https://github.com/lowdefy/lowdefy/assets/24702285/b3c5549d-2920-4460-8023-f35f01762cea)

![image](https://github.com/lowdefy/lowdefy/assets/24702285/bf45fab1-bd7e-4c3d-906c-d69cb34b078c)

This means that (at least on Windows), for this to work with relative paths (I suspect that this issue would affect other areas), you'll need to either change the working directory to where the yaml files. Or allow the user to define where the node/lowdefy files are and where the app yaml files (which would probably be a more versatile option).

**Reply by SamTolmay** [reply]
I had forgotten, but this is why we recommended using absolute paths for the filename. The server, and therefore also the SQLite driver are running from the server directory, and therefore it interprets paths relative to that directory, not the config directory (where the lowdefy.yaml file is placed).

We download the server code to the `./.lowdefy/dev` and `./.lowdefy/server` directories to avoid clashes with other files. You can configure these directories by using the `--config-directory`, `--dev-directory` and `--server-directory` CLI arguments, or the `LOWDEFY_DIRECTORY_CONFIG`, `LOWDEFY_DIRECTORY_DEV` and `LOWDEFY_DIRECTORY_SERVER` environment variables (https://docs.lowdefy.com/cli).

Because of the way these directories are structured, you should be able to get things to work if you specify the file path relative to the server directory - i.e. if you do `../../tutorial_db.sqlite` instead of `./tutorial_db.sqlite`. This is a little confusing for new users, which is why we recommended the absolute path, but maybe it makes sense to have a deeper understanding of whats happening under the hood.
